### DISCUSSION TITLE
public/api/timesheets [403] Forbidden error

### DISCUSSION THREAD

**Discussion by hyssop** [kimai/kimai#3950] [On-Premise: Installation & Updates] [answered]
I have Kimai 1.30.11 installed via Softaculous on a cPanel server. Everything is fine except that api calls seem to fail for a permission issue. I tried manual (git/composer) install and got the same results The server is CentOS/Apache/MariahDB/PHP-FPM with nGinx pre-proxy. Everything is at the latest build. I have enabled the last 2 lines in .env. Kimai Doctor shows all green (perms, extensions, no log file).

Examples of the error are trying to stop the timer, or restart an activity, or delete an entry in My Times. They result in: "Time recording could not be started [403] Forbidden" or similar wording.

I haven't spent a lot of time searching through the code to track down the source of the error and thus it's cause. /pulbic/en, /public/api, appear to be redirects since the folders don't exist, but whatever public/api/ is trying to do it failing.

Has anyone resolved this problem?
Logs to check?
Permissions to check? 

-Pete

![Screenshot 2023-03-27 033854](https://user-images.githubusercontent.com/7874162/228124079-eed96d40-bae0-42b4-af6b-c231d94b7385.png)

**Comment by kevinpapst** [comment]
This happens when you use the wrong HTTP method. Is some plugin blocking PATCH requests or is the JS broken?
The links that are triggered are converted via JS into PATCH requests to the API. If you open them via GET, you will see this error.

Yes, that is a Softaculous issue... butut: Kimai never officially supported to run in a subdirectory. If you see `public/` in the URL you are using it wrong. DocumentRoot needs to point to public directly, so it does not appear in the URL.

**Reply by softaculous** [reply]
Hi,

I am from the Softaculous team.

@kevinpapst PATCH request not working indicates that there is a configuration issue with the web server i.e. Apache/NGINX which is not managed by Softaculous but is managed by the control panel in this case cPanel. So I can confirm that if the above error is due to PATCH request not working it is NOT a Softaculous issue. 

Regarding the document root, cPanel and other control panels do not allow changing the document root of the primary domain. So you cannot point your domain directly to public/ folder. Hence when installing Kimai we create a .htaccess file which redirects all requests made to the domain to /public

Hope this makes things clear.

**Reply by hyssop** [reply]
@softaculous  Thank you for your input here, and on the Softaculous support ticket. There are two issues I was facing: the document root needing to point to /public, and PATCH not being allowed.

On the first issue, I was unaware of the Softaculous workaround via .htaccess, but I see it now. It still allows /public/ to show in the URL (e.g. kimai.example.com/public/en/timesheet/), but that is expected, I assume. 

Another option in cPanel is to install and then change documentroot, adding /public to it, but that I expect might break some things (updates? backups?) for Softaculous. The Kimai install page (https://www.kimai.org/documentation/installation.html) says "Configure your web server (like Nginx or Apache) to point its DocumentRoot at the public/ directory. For more details, see the Webserver How-To." I missed this line the first time through. So the manual install process makes sense now.

The second issue is indeed a server config issue. It's odd to me that both the cPanel servers I tested don't allow PATCH, but I will resolve that with cPanel if I can't sort it on my own.

@kevinpapst Thank you for you patient responses. Kimai is a great alternative to Clockify! :)

**Reply by kevinpapst** [reply]
Hahaha 😁 You know, it is hard to handle support requests (on a daily basis) which are in > 70% based on the user environment. I totally understand why people then come here, but each and everyone is taking away a tiny bit of time of my life... So I am a lot, but usually not patient. 

On the topic: I have had support requests about such problems, but the last one was a long time ago. When PATCH requests are blocked it was usually an ancient proxy server in between the user and Kimai that did not pass on the PATCH requests. Which is wrong on so many levels... but anyway. I have not yet seen a default Apache/nginx config that would disallow PATCH requests. Maybe some mentally ill "security plugin" that is a bit too strict?

From my experience CPanel / Softaculous installations simply work (even though not using public/ as document root).

Did you try curl to make sure that this is not a browser issue?

**Reply by hyssop** [reply]
> ...it is hard to handle support requests (on a daily basis) which are in > 70% based on the user environment. I totally understand why people then come here, but each and everyone is taking away a tiny bit of time...
> 
I get it! I do a lot of end user support; it can be **SO** time consuming! 😉 

The user environment has nearly endless variable combinations, and there's always one more weird situation waiting out there.
> I have not yet seen a default Apache/nginx config that would disallow PATCH requests. Maybe some mentally ill "security plugin" that is a bit too strict?
> 
LOL, I agree! PATCH should just work on Apache/nGinx. Now that I know that it's not related to installation issues (location of doc root, etc.) I'll be able to find the problem. I've seen times when ModSec had to be tweaked to allow something, but again, I wouldn't expect cPanel to set up a server to block PATCH by default.
> From my experience CPanel / Softaculous installations simply work (even though not using public/ as document root).
> 
Same here!
> Did you try curl to make sure that this is not a browser issue?
>
I confirmed that your demo works for me, so I know it's not anything between my keyboard and the server. It's definitely an on-server issue. Some security tweak, I'll bet. I'll find it. 🤓

**Comment by hyssop** [comment]
Thank you for explaining the process with the links. I assumed something of that nature, that GET was the 403 issue. I don't think it is being blocked, I think it has to do with the document root issue, or at least I should start there. But I have questions...

Before posting here, I was reading more about the installation, and that Kimai is not supported for subdirectory install, so I tried the CLI installation. In your docs you say "connect with SSH to your server and change to your webservers (document) root directory." From there I assumed we are to run the commands. I followed this (from https://www.kimai.org/documentation/installation.html) and it resulted in the same file structure as the Softaculous install:

/public_html/kimai/public/

"If you see public/ in the URL..." Yes, I thought that was strange also, but seemed to be the result either way I installed it.

Did I miss something, or does the web document root need to be changed AFTER the installation?

-Pete

**Reply by kevinpapst** [reply]
Document root of your webserver has nothing to do with your SSH structure or the public_html directory (whatever that is): https://www.kimai.org/documentation/webserver-configuration.html

**Reply by kevinpapst** [reply]
> I assumed something of that nature, that GET was the 403 issue. I don't think it is being blocked, I think it has to do with the document root issue, or at least I should start there. But I have questions...

One more time: PATCH vs GET. Check if your browser is really sending PATCH request. Do so by opening your browser console and then verifying the requests in the network panel.

**Reply by hyssop** [reply]
> > I assumed something of that nature, that GET was the 403 issue. I don't think it is being blocked, I think it has to do with the document root issue, or at least I should start there. But I have questions...
> 
I was saying that GET was a problem (understood that PATCH is the way).
> One more time: PATCH vs GET. Check if your browser is really sending PATCH request. Do so by opening your browser console and then verifying the requests in the network panel.

Yes, they are PATCH (or DELETE). But the response is HTTP/1.1 405 (403 for DELETE), so I need to look into why that is. PATCH works in HTML/1.1 afaik. Kimai doesn't require HTTP/2, correct?

**Comment by hyssop** [comment]
public_html is a very common webserver documentroot. You are correct that is has nothing to do with SSH, nor did I say that. Your installation instructions say to SSH into and install from the webserver documentroot which yields the result of Kimai being installed in a subdirectory. Not trying to be rude, but I think the wording on the installation instructions could be more clear, and that would help.

Oddly, Softaculous claims that their installer works for them (without the 403 error) but I don't see how! From what you tell me it shouldn't work. I am asking them for clarification.

I found https://www.kimai.org/documentation/webserver-configuration.html just after I made my last post and from that determined the you do intend your /public/ folder to be the webserver documentroot. That's easily changed and I did so. Now the URLs look the same as your demo site. Kimai basically works, as it did prior to the change, but I am still getting /api/ errors as follows:

Stopping timer: "Time recording could not be stopped
Method Not Allowed"

Restarting an activity: "Time recording could not be started
Method Not Allowed"

Deleting events: "Entry could not be deleted:
[403] Forbidden"

I have no plugins installed. I will keep looking in docs for some perms issue, or something, but any pointers are appreciated.

-Pete

**Comment by hyssop** [comment, marked as answer]
I'm finally getting back to this...

As expected, the problem is not with Kimai (nor Saftaculous) but is entirely a ModSecurity issue. I can't come up with a universal solution, due to variations in OWASP rule sets which may be in use, but apparently by default HTTP methods PATCH and DELETE are denied, and they are necessary for Kimai's functions.

**The fix is a simple matter of finding and duplicating the HTTP method authorization rule (search rules for "allowed_methods"), disable the original rule, then add the needed methods to the duplicate rule.**

Since cPanel with ModSecurity is a huge portion of both Kimai's and Softaculous' installation base, it would be advisable to add notes to users about this issue.

Thanks again for all the information that helped me sort this out.

-Pete

**Reply by softaculous** [reply]
Glad to know you were able to find the exact cause of the issue.
